fish-rust/src/bitset.rs

```diff
@@ -0,0 +1,275 @@
+use num_traits::bounds::UpperBounded;
+use num_traits::{AsPrimitive, FromPrimitive, Num};
+use std::ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign, Not, Shl, Shr};
+
+pub struct BitSet<T>(T);
+
+impl<
+        T: BitAnd<T, Output = T>
+            + BitOr<T, Output = T>
+            + BitAndAssign<T>
+            + BitOrAssign<T>
+            + Shl<usize, Output = T>
+            + Shr<usize, Output = T>
+            + Not<Output = T>
+            + AsPrimitive<u8>
+            + FromPrimitive
+            + UpperBounded
+            + Num
+            + AsI64
+            + AsUsize,
+    > BitSet<T>
+where
+    i64: AsPrimitive<T>,
+{
+    pub fn set(&mut self, i: usize) {
+        self.0 |= T::one() << i;
+    }
+
+    pub fn unset(&mut self, i: usize) {
+        self.0 &= !(T::one() << i);
+    }
+
+    pub fn toggle(&mut self, i: usize, v: bool) {
+        // if v { self.set(i) } else { self.unset(i) }
```

Comment:
leftover comment

Reply:
It's intentional to clarify the logical equivalent. Maybe I should another line saying that.

Reply:
Could put it in a doc comment instead, maybe even with an example (which becomes a doctest)

Reply:
Good idea. doctests aren't working atm (I think because they're all private modules).